DESIGN.md

```diff
@@ -0,0 +1,773 @@
+designing the Rust APIs using ownership passing, Rust enforces the requirements
+at compile time. The following example demonstrates reading and writing with a
+file resource.
+
+```rust
+use tokio_uring::buf;
+
+/// The result of an operation that includes a buffer. The buffer must
+/// be returned to the caller when the operation completes successfully
+/// or fails.
+///
+/// This is implemented as a new type to implement std::ops::Try once
+/// the trait is stabilized.
+type BufResult<T> = (std::io::Result<T>, buf::Slice);
+
+// Positional read and write function definitions
+impl File {
+    async fn read_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+    
+    async fn write_at(&self, buf: buf::Slice, pos: u64) -> BufResult<usize>;
+}
+
+/// The caller allocates a buffer
+let buf = buf::IoBuf::with_capacity(4096);
+
+// Read the first 1024 bytes of the file, when `std::ops::Try` is
+// stable, the `?` can be applied directly on the `BufResult` type.
+let BufResult(res, buf) = file.read_at(0, buf.slice(0..1024)).await;
+// Check the result.
+res?;
+
+// Write some data back to the file.
+let BufResult(res, _) = file.write_at(1024, buf.slice(0..512)).await;
+res?;
+```
+
+When reading from a TCP stream, read operations remain in flight until the
+socket receives data, an exchange that can take an arbitrary amount of time.
+Suppose each read operation requires pre-allocating memory for the in-flight
+operation. In that case, the amount of memory consumed by the process grows
+linearly with the number of in-flight read operations. For applications with a
+large number of open connections, this can be problematic. The io-uring API
+supports registering buffer pools with the ring and configuring read operations
+to use the buffer pool instead of a dedicated per-operation buffer. When the
+socket receives data, the kernel checks out a buffer from the pool and returns
+it to the caller. After reading the data, the caller returns the buffer to the
+kernel. If no buffers are available, the operation pauses until the process
+returns a buffer to the kernel, at which time the operation completes.
+
+```rust
+// ...
+let buf = file.read_at_prepared(0, 1024, DefaultPool).await?;
+
+// Write some data back to the file.
+let BufResult(res, buf) = file.write_at(1024, buf.slice(0..512)).await;
+res?;
+
+// Dropping the buffer lets the kernel know the buffer may
+// be reused.
+drop(buf);
+```
+
+The runtime supports initializing multiple pools containing buffers of different
+sizes. When submitting a read operation, the caller specifies the pool from
+which to draw a buffer.
+
+```rust
+// Allocate a buffer pool
+let my_pool = BufferPool::builder()
+    .buffer_size(16_384)
+    .num_buffers(256)
+    .build();
+
+// Create the runtime
+let mut rt = tokio_uring::runtime::Runtime::new()?;
+
+// Provide the buffer pool to the kernel. This passes
+// ownership of the pool to the kernel.
+let pool_token = rt.provide_buffers(my_pool)?;
```

Comment:
Maybe the runtime should have a method for taking a list of pools, something like

```rust
let mut rt = tokio_uring::runtime::Runtime::new().
    .with_pools([netpool, fspool,...])
    .build()?;
```

Or

```rust
let mut rt = tokio_uring::runtime::Runtime::new().
    .with_pool(netpool)
    .with_pool(fspool)
    //...
    .build()?;
```

Reply:
There needs to be a way to identify a pool (i.e. the buffer group in io-uring API) to read operations that use it.